docs/design/proposals/direct-blk-device-assignment.md

```diff
@@ -0,0 +1,264 @@
+```Golang
+type NodeGetVolumeStats struct {
+	// This field is OPTIONAL.
+	Usage []*VolumeUsage `json:"usage,omitempty"`
+	// Information about the current condition of the volume.
+	// This field is OPTIONAL.
+	// This field MUST be specified if the VOLUME_CONDITION node
+	// capability is supported.
+	VolumeCondition      *VolumeCondition `json:"volume_condition,omitempty"`
+}
+```
+
+#### Quick notes on implementation
+
+Instead of the CSI node driver writing the mount info into a `csiPlugin.json` file under the volume root, 
+as described in the original proposal. Here we propose that the CSI node driver passes the mount information to 
+the kata-runtime through a new kata-runtime command called `add-blk-mount-info`. The kata-runtime then write the mount 
+information to a `“`mount-info.json`”` file in a hardcoded location (e.g. `/run/kata-containers/shared/direct-block-devices/[block_dev_name]/`).
+
+When a Kata Pod sandbox is started, the containerd-kata-shim-v2 checks whether there is any mount identified 
+by its source path in the container specification that has a mount info file in the `direct-block-devices` directory. 
+If it exists, the containerd-kata-shim-v2 parses the mount info file and propagates it to the kata agent to carry out the mount. 
+Lastly,the containerd-kata-shim-v2 creates a file named by the sandbox id under the `direct-block-devices/[block_dev_name]/` 
+directory. The reason for adding a sandbox id file is to establish a mapping between the volume and the sandbox. 
+Later, when the kata-runtime handles the `get-stats` and `resize` commands, it can go to the volume’s directory, 
+find the sandboxes that mount this volume, and then identify the endpoint of the corresponding containerd-shim-kata-v2.
+
+### containerd-shim-kata-v2 changes
+containerd-shim-kata-v2 provides an API for sandbox management through a unix domain socket. Two new handlers are proposed:
+```
+New handlers:
+/fs-stats
+/fs-resize
+
+Example:
+
+curl —unix-socket <shim socket path> -I \
+-X GET ‘http://localhost/fs-stats' \
+-d ‘{ “block-device”: “/dev/sda” }’ 
+
+curl —unix-socket <shim socket path> -I \
+-X _PUT_ ‘http://localhost/fs-resize' \
+-d ‘{ “block-device”: “/dev/sda”,
+“volume-size”: “123123” 
+}’
+```
+
+The shim will then forward the corresponding request to the kata-agent to mount the filesystem, 
+gather the stats and refresh the filesystem metadata. For `fs-resize` operation, 
+kata-runtime might also need to notify the hypervisor to resize the block device (e.g. call `block_resize` in QEMU). 
+
+### Kata-agent changes
+
+For mounting a filesystem on the block device volume, the information can be passed to Kata-agent through the existing 
+`Storage` grpc object. Two new grpc request objects are added to the Kata-agent grpc interface for 
+resizing and getting volume stats:
+```Golang
+type ResizeFileSystemRequest struct {
+  mountPoint string
+  size       int64
```

Comment:
In the example above you've specified an actual (new) volume side so this isn't a delta value. In other words, if a volume was _reduced_ in size, the `size`/`volume-size` would still be positive right? In which case, why is this not `uint64`?

Also, please specify the units for this struct member.